### DISCUSSION TITLE
How can I write a "site-block" userscript to let me block a list of (distracting) sites as required?

### DISCUSSION THREAD

**Discussion by sanjayankur31** [qutebrowser/qutebrowser#6726] [Q&A] [answered]
I'd like to block a list of sites during specific times, working hours for example. It's not meant to be a privacy thing, just to prevent me from distracting myself over news sites and so on (which are so easy to open when I'm waiting for some thing to finish, for example).

My first attempt was to use the `~/.config/qutebrowser/blocked-hosts` file. I made a list of these sites, and with a shell script, I could rename it to `~/.config/qutebrowser/blocked-hosts` when I want to block these sites:

```
#!/bin/bash

enable_block () {
    pushd ~/.config/qutebrowser
        if [ -f "blocked-hosts" ]
        then
            echo "OK: host blocking already enabled"
            exit 0
        fi

        if [ -f "blocked-hosts.txt" ]
        then
            cp blocked-hosts.txt blocked-hosts && echo "OK: host blocking enabled"
        else
            echo "ERROR: blocked-hosts.txt not found."
            exit -1
        fi
    popd
}
disable_block () {
    pushd ~/.config/qutebrowser
        if [ -f "blocked-hosts" ]
        then
            rm -f blocked-hosts && echo "OK: host blocking disabled"
        else
            echo "OK: host blocking already disabled"
            exit 0
        fi
    popd
}

toggle_block () {
    pushd ~/.config/qutebrowser
        if [ -f "blocked-hosts" ]
        then
            disable_block
        else
            enable_block
        fi
    popd
    exit 0
}

function usage() {
    echo "$0: Update user host blocking list"
    echo
    echo "Usage: $0 [-edth]"
    echo
    echo "-e: enable host blocking list"
    echo "-d: disable host blocking list"
    echo "-t: toggle host blocking"
    echo "-h: print this usage text and exit"
}

if [ $# -lt 1 ]
then
    usage
    exit 1
fi

# parse options
while getopts "edth" OPTION
do
    case $OPTION in
        e)
            enable_block
            exit 0
            ;;
        d)
            disable_block
            exit 0
            ;;
        t)
            toggle_block
            exit 0
            ;;
        h)
            usage
            exit 0
            ;;
    esac
done
```

This mostly works, but I've noticed that sometimes even after I enable/disable, qutebrowser doesn't immediately "pick up" the settings. 

What would be the suggested way of implementing this sort of functionality in qutebrowser?

**Comment by wolass** [comment]
I agree that this functionality would be very much appreciated!!! I get disctracted from writing papers all the time... 

So looking at your code just a noob sugestion:

Arch wiki says: 

```
Create ~/.config/qutebrowser/blocked-hosts and 
enter websites you want to block in each line; www.youtube.com for example. 
This will keep the built-in adblock list while adding the websites in. 
Restart qutebrowser, and 
run :adblock-update.
```

What is your command in qutebrowser that is bound to a key? 
Do you update adblock each time you toggle distracting hosts?

Is the behaviour consistent if you restart the qutebrowser?

**Comment by wolass** [comment, marked as answer]
So basically I think that QB writes the blocked hosts to some other file in a current session. 

But after saving your script in a distractor_blocker file and making it executable 
and putting this into my config.py : 

```
config.bind('b', 'spawn bash ~/.config/qutebrowser/distractor_blocker -t ;; adblock-update ;; restart')
```
I get the expected behaviour which seems consistent,

The only downside is that it has to restart, so I messes up my window manager tab-association.
Let me know if there is a better way.

**Reply by sanjayankur31** [reply]
Yeh, the restarting bit is quite a limitation for me too---I generally have about 5 windows spread over different workspaces etc., and it's just an extra step to reorganise them all etc.

@The-Compiler : would there be a way of modifying the block list dynamically, without restarting qutebrowser? If there is, that'll be ideal and we could write a userscript around it :)

**Reply by toofar** [reply]
You shouldn't need to restart the browser. Editing `~/.config/qutebrowser/blocked-hosts` then running `:adblock-update` should do it. If it doesn't seem to be working when you run `:spawn some_script ;; adblock-update` it is probably because `;;` doesn't block the next command until the previous one finishes, it'll run them both as soon as possible. So the block list might be being read before the script has updated it.
I would suggest turning your script into a userscript so that it can control when ablock-update is called itself. So the userscript would change the file then write the update command to the fifo.
See https://qutebrowser.org/doc/userscripts.html and some examples in https://github.com/qutebrowser/qutebrowser/tree/master/misc/userscripts

**Reply by sanjayankur31** [reply]
So enabling the block works for me, but not disabling. I.e., if I remove the `blocked-hosts` file and then run `:adblock-update`, the hostnames that were in the `blocked-hosts` files are still blocked. So it seems to add new hosts to the "in memory" list fine, but doesn't remove them? As @wolass noted, restarting is the only reliable way to ensure that this works.

**Reply by sanjayankur31** [reply]
PS: this is easy to convert to  a userscript, but I'll do that once we've worked all the kinks out :)

**Reply by toofar** [reply]
Right, I understand what the issue is now. The hostblocker has two internal variables: `_blocked_hosts` which holds the compiled set of hosts from the subscribed lists, and `_config_blocked_hosts` which stores the hosts from ~/.config. When you run `:adblock-update` then `_blocked_hosts` is [cleared](https://github.com/qutebrowser/qutebrowser/blob/57f24e7/qutebrowser/components/hostblock.py#L217) before re-processing but `_config_blocked_hosts` isn't. So the one in ~/.config is only additive.
I'm pretty sure that is a bug.

**Reply by The-Compiler** [reply]
Yep, that looks like a bug indeed, I opened #6753. Thanks @toofar for tracking this down!

@sanjayankur31 If there's nothing urgent or something you're sure I have to look at personally, I'd prefer not being mentioned - I look through the backlog for all of qutebrowser's communication channels (issues, PRs, discussions, Reddit, IRC, mailinglist, personal mail, etc.), but I'd rather do so on my own schedule. As you've seen, often there are other people who can help too :wink:

**Reply by sanjayankur31** [reply]
Sure thing.
